File: Examples/Python/underautomation/universal_robots/urarchive.py
import typing
import clr
import os
clr.AddReference(os.path.realpath(os.path.join(os.path.dirname(__file__), "..", "..", 'lib', 'UnderAutomation.UniversalRobots.dll')))
from UnderAutomation.UniversalRobots import URArchive as urarchive

class URArchive:
	def __init__(self, _internal = 0):
		if(_internal == 0):
			self._instance = urarchive()
		else:
			self._instance = _internal
	def save(self, stream: typing.Any) -> None:
		self._instance.Save(stream)
	def save(self, directory: str) -> str:
		return self._instance.Save(directory)
	@staticmethod
	def load(filePath: str) -> typing.Any:
		return urarchive.Load(filePath)
	@staticmethod
	def load(fileStream: typing.Any) -> typing.Any:
		return urarchive.Load(fileStream)
	@property
	def xml(self) -> typing.Any:
		return self._instance.XML
	@property
	def name(self) -> str:
		return self._instance.Name
	@name.setter
	def name(self, value: str):
		self._instance.Name = value
	@property
	def file_name(self) -> str:
		return self._instance.FileName

File: UnderAutomation.UniversalRobots.ObfuscatedSources/Files/URArchive.cs
//              WARNING
// This file is an empty shell containing only public C# items.
// The internal code is hidden; to access it, you need to obtain a Source licence of the library.

using System.Xml.Linq;
using System.IO;

namespace Files {
	/// <summary>
	/// Contains basic methods to encode and decode a UR archive
	/// </summary>
	public abstract class URArchive {

		/// <summary>
		/// Initializes a new instance of <xref href="UnderAutomation.UniversalRobots.Files.URArchive" data-throw-if-not-resolved="false"></xref> from the specified XML element.
		/// </summary>
		/// <param name="xml">The XML element representing the archive content.</param>
		protected URArchive(XElement xml)
		{
			// Source is hidden, a Source licence is needed to access internal code...
		}

		/// <summary>
		/// Save encoded file to a stream
		/// </summary>
		public void Save(Stream stream)
		{
			// Source is hidden, a Source licence is needed to access internal code...
		}

		/// <summary>
		/// Save encoded file to a directory, overwrite it if it exists
		/// </summary>
		/// <param name="directory">Directory to save the file</param>
		/// <returns>Full name of the saved file</returns>
		public string Save(string directory)
		{
			// Source is hidden, a Source licence is needed to access internal code...
			return default;
		}

		/// <summary>
		/// Load a UR archive from file path and decode it as XML
		/// </summary>
		public static XElement Load(string filePath)
		{
			// Source is hidden, a Source licence is needed to access internal code...
			return default;
		}

		/// <summary>
		/// Load a UR archive from stream and decode it as XML
		/// </summary>
		public static XElement Load(Stream fileStream)
		{
			// Source is hidden, a Source licence is needed to access internal code...
			return default;
		}

		/// <summary>
		/// XML description of the object
		/// </summary>
		public XElement XML { get; }

		/// <summary>
		/// Gets the XML attribute name used to store the archive name.
		/// </summary>
		protected abstract string NameAttribute { get; }

		/// <summary>
		/// Gets the expected root XML element name for this archive type.
		/// </summary>
		protected abstract string RootElement { get; }

		/// <summary>
		/// Gets the file extension for this archive type (e.g., ".urp" or ".installation").
		/// </summary>
		protected abstract string Extension { get; }

		/// <summary>
		/// Gets or sets the name of this archive, stored as an XML attribute.
		/// </summary>
		public string Name { get; set; }

		/// <summary>
		/// File name that should be used on a UR robot
		/// </summary>
		public string FileName { get; }
	}
}
